csvcubed/tests/unit/cli/build/test_build_with_config.py

```diff
@@ -0,0 +1,626 @@
+
+PROJECT_ROOT = Path(Path(__file__).parent, "..", "..", "..", "..").resolve()
+TEST_CASE_DIR = Path(get_test_cases_dir().absolute(), 'cli', 'build')
+SCHEMA_PATH_FILE = Path(PROJECT_ROOT, "csvcubed", "schema", "cube-config-schema.json")
+
+
+def test_01_01_build_ok():
+    config = Path(TEST_CASE_DIR, "cube_data_config_ok.json")
+    output = Path("./out")
+    csv = Path(TEST_CASE_DIR, "cube_data_config_ok.csv")
+    cube, validation_errors = cli_build(
+        config_path=config,
+        output_directory=output,
+        csv_path=csv,
+        fail_when_validation_error_occurs=False,
+        validation_errors_file_out=Path("validation_errors.json"),
+    )
+    assert isinstance(cube, Cube)
+    assert isinstance(validation_errors, list)
+    assert len(validation_errors) == 0
+
+
+def test_01_02_build_ok_fail_on_error():
+    config = Path(TEST_CASE_DIR, "cube_data_config_ok.json")
+    output = Path("./out")
+    csv = Path(TEST_CASE_DIR, "cube_data_config_ok.csv")
+    cube, validation_errors = cli_build(
+        config_path=config,
+        output_directory=output,
+        csv_path=csv,
+        fail_when_validation_error_occurs=True,
+        validation_errors_file_out=Path("validation_errors.json"),
+    )
+    assert isinstance(cube, Cube)
+    assert isinstance(validation_errors, list)
+    assert len(validation_errors) == 0
+
+
+def test_01_03_build_exit_on_fail():
+    """
+    Tests that the code exits on failure rather than returning incomplete cube with errors
+    """
+    config = Path(TEST_CASE_DIR, "cube_data_test_02_01.json")
+    output = Path("./out")
+    csv = Path(TEST_CASE_DIR, "cube_data_test_02_01.csv")
+    with pytest.raises(SystemExit) as err:
+        cube, validation_errors = cli_build(
+            config_path=config,
+            output_directory=output,
+            csv_path=csv,
+            fail_when_validation_error_occurs=True,
+            validation_errors_file_out=Path("validation_errors.json"),
+        )
+    assert err.value.code == 1
+
+
+def test_02_01_val_errors_no_observation():
+    """
+    Test for:-
+    NoObservedValuesColumnDefinedError
+    """
+    config = Path(TEST_CASE_DIR, "cube_data_test_02_01.json")
+    output = Path("./out")
+    csv = Path(TEST_CASE_DIR, "cube_data_test_02_01.csv")
+    cube, validation_errors = cli_build(
+        config_path=config,
+        output_directory=output,
+        csv_path=csv,
+        fail_when_validation_error_occurs=False,
+        validation_errors_file_out=Path("validation_errors.json"),
+    )
+    # Check cube
+    assert isinstance(cube, Cube)
+    assert isinstance(cube.columns, list)
+    assert len(cube.columns) == 3
+
+    # Check returned errors
+    assert isinstance(validation_errors, list)
+    assert isinstance(validation_errors[0], NoObservedValuesColumnDefinedError)
+
+    # Check logged validation errors
+    log_filepath = Path(appdirs.user_log_dir(), '..', 'csvcubed', 'csvcubed_testing', 'Logs',
```

Comment:
haven't checked them all, but you can probably assign a constant at the top of the file and just reuse throughout.

Reply:
Created Constant and extracted log read to function.